### ISSUE TITLE
Course Material & News

### ISSUE THREAD

**Issue by ningyuansg** [source-academy/cadet#52]
### Summary

In frontend's [#81](https://github.com/source-academy/cadet-frontend/issues/81), we agreed with Prof @martin-henz to move the material (previously 'materials') and news (previously 'announcments') content to public access.

On the sprint 2 review meeting (9th June), there was some discussion about outsourcing the management of material and news to NUS IVLE. Source academy's material will map to IVLE's files (workbin); and likewise news will map to IVLE announcements.

This means that we can let IVLE handle user uploads, file and post management. **Course administrators on IVLE will also be able to upload a file or create an announcement once, and have it appear on both platforms.**

### Proposed Solution

However, for this to work well, we will need IVLE to either
1. Provide a dummy IVLE user account with its own API key, or
2. Provide an API key that simply allows access to the CS1101S module on IVLE, or
3. Make the CS1101S module on IVLE's announcements and files public

A dummy IVLE user account can be added as a student to the CS1101S module on IVLE. We can then access the IVLE announcements and files (workbin) through this dummy account. As long as the dummy account is in each semester's CS1101S module, we will not need to change anything from semester to semester.

An API key that directly provides access to the CS1101S module would make things much simpler, but I suspect that IVLE does not have the ability to generate such a key. Similarly, option 3 of making course announcements and files public is the simplest way out, but again IVLE may not have this capability. 

To conclude, we should request for option 3 first, then 2, then 1.

### Backend

Should option 3 be made possible for us, there would be no need for the backend code base to act.

For options 1 and 2, however, my opinion is that the frontend should not fetch IVLE data directly. Instead, the backend should periodically fetch and store the information from IVLE, and then provide it to the frontend via API call. This ensures that neither the API key nor the user authentication token (for option 1) will be exposed to the end-user.

**Comment by ningyuansg** [comment]
From today's meeting with the SoC systems engineer:

### On topic

- We are able to use a 'dummy account' by means of the 'guest account' feature available Prof @martin-henz on IVLE in order to access IVLE announcements, files (workbin), and possibly webcasts. That is, we will probably be implementing solution (1) above, upon confirmation by the backend team.
- If this guest account can request for it's own API key, then we are able to use that API key as our dedicated production API key. This is currently being checked, and the systems engineer will get back to us.
- The systems engineer will also verify for us if there is a cap to API calls per key. If there is, we should request to be put on a whitelist to accommodate smooth registration of students at the start of semester 1.

### The issue

The issue, specifically, is to implement the fetching, storage, and provision via API calls of IVLE announcements, files (workbin), and perhaps webcasts. 

### Off topic

- Security concerns were raised with regards to the 52 weeks validity of the refresh token. (#53)
- There is a need to add one more role for graduate students. (#54)

**Comment by ningyuansg** [comment]
**Diagram of how to programmatically obtain authentication tokens for our guest account:**

![photo6152448550697543734](https://user-images.githubusercontent.com/22320312/41770994-57cb2bb0-7646-11e8-9589-c077c892bf45.jpg)

**Python3 script implementing the above diagram**

```python3
#! /usr/bin/env python3

import bs4, getpass, re, requests, urllib.parse

KEY = input('API key (https://ivle.nus.edu.sg/LAPI/default.aspx): ')
LOGIN_URL = 'https://ivle.nus.edu.sg/api/login/?apikey={}&url=localhost'.format(KEY)
LOGIN_PAYLOAD = {
    'userid': input('NUSNET ID: '),
    'password': getpass.getpass('Password: ')
}

session = requests.session()

response_viewstate = session.get(LOGIN_URL)
soup = bs4.BeautifulSoup(response_viewstate.text, 'html.parser')
__VIEWSTATE = soup.find('input', {'id': '__VIEWSTATE'}).get('value')

LOGIN_PAYLOAD['__VIEWSTATE'] = __VIEWSTATE

response_post = session.post(LOGIN_URL, data=LOGIN_PAYLOAD)
url_with_token = response_post.url
token_query = urllib.parse.urlparse(url_with_token).query
print(re.search('(?<=token=).+\Z', token_query).group())
```

(You will need to install the _requests_ and _bs4_ libraries. `$ pip3 install requests beautifulsoup4`. The rest of the libraries are part of the standard python3 library)

**Comment by ningyuansg** [comment]
I'd say most of this was fixed in #72, and the rest is covered by #79.
